packages/next/src/server/request/params.ts

```diff
@@ -93,7 +94,7 @@ export function createPrerenderParamsForClientSegment(
   workStore: WorkStore
 ): Promise<Params> {
   const prerenderStore = prerenderAsyncStorage.getStore()
-  if (prerenderStore) {
+  if (prerenderStore && prerenderStore.type === 'prerender') {
     if (isDynamicIOPrerender(prerenderStore)) {
```

Comment:
You can omit the extra type check and just call isDynamicIOPrerender in the outer scope since the else case falls through to the same result regardless

Reply:
It’s needed in various places because the type needs to be refined too. Typescript doesn’t know that isDynamicIOPrerender is the modern prerender.

However I’m also writing in a style where I expect these stores to be unified and so it’s actually refining from the larger WorkUnit too.

Reply:
`isDynamicIOPrerender` takes an unrefined store so the extra refinement here is duplicative. Additionally we can add

```
function isDynamicIOPrerender(store: PrerenderStore): store is PrerenderStoreModern {
  ...
}
```
To get this function to do the refinement for us.

Probably should add a

```
function isModernPrerenderStore(store: undefined | PrerenderStore): store is PrerenderStoreModern {
  return store && store.type === 'prerender'
}
```
and replace all the inline type checks too since the inlined code doesn't convey the intent (it's not obvious that there is a prerender-legacy type that you are trying to exclude